code/game/machinery/doors/door.dm

```diff
@@ -67,8 +80,45 @@
 	health = maxhealth
 
 	update_nearby_tiles(need_rebuild=1)
+	if (hashatch)
+		setup_hatch()
 	return
 
+/obj/machinery/door/proc/setup_hatch()
+
+	if (overlays != null)
+		hatch_image = image('icons/obj/doors/hatches.dmi', src, hatchstyle, closed_layer+0.1)
+		hatch_image.color = hatch_colour
+		hatch_image.pixel_x = hatch_offset_x
+		hatch_image.pixel_y = hatch_offset_y
+
+		overlays += hatch_image
+		update_icon()
+	else
+		spawn(10)
+			setup_hatch()
+
+/obj/machinery/door/proc/open_hatch(var/atom/mover = null)
+	if (!hatchstate)
+		hatchstate = 1
+		update_icon()
+		playsound(src.loc, hatch_open_sound, 40, 1, -1)
+	hatchclosetime = world.time + 29
+
+	if (istype(mover, /mob/living/silicon))
+		var/mob/living/silicon/S = mover
+		S.under_door()
+	spawn(30)

```

Comment:
This is dangerous.

Either have it execute right after the `spawn()` command, _without_ the if check right afterwards. Or tie it to the `process()` proc in the following manner:
```
if (hatchstate && world-time > hatchclosetime)
    close_hatch()
```

Do not use two methods of timing, on the very small off-chance that they might somehow desynchronize under strain.